### ISSUE TITLE
why do we have negative self-attention score?

### ISSUE THREAD

**Issue by muiPomeranian** [lucidrains/reformer-pytorch#101]
hey lucid ! sorry to ask this dumb question .. 

so what I understand about self attention is it gives the relative information among all words(representing relevance of word?). 

So I expected to see non negative value but apparently there were many negative self-attention score inside since we initialized the Q,K from uniform -0.01~0.01 in the beginning. 

1)
What is the math or reasoning behind that we initialize all learnable params with uniform -.01~0.01? do you intend to assume that we are living in sphere(geometrical reasoninng)? 

2) 
how would you interpret the negative val of self attention score? 

3) 
I think this will depend on your thought/advice for my Q2, but I am thinking of assuming that if I have negative val, it means  it is MORE Negatively important as for the relevance info. which means, like a "BAD TO HAVE" ? then how would you interprete self-attention score of 0? I would guess it is not important at all?

Thanks again a lot...really...

**Comment by lucidrains** [comment]
@muiPomeranian I don't believe I have any initialization code, at least, in the most recent version. I leave it up to the user, since I have not found a major difference in training, compared to, say, placement of layer normalization.

I think the attention values you should be looking at are not after the dot product of the queries and keys, but after it is masked and then normalized by the softmax (bringing it into a range of 0-1) For each layer, the attention values are interpreted as the relative importance each token places on all other tokens.

**Comment by muiPomeranian** [comment]
Thanks for the response again !!

But I found out that we don't just attend within same bucket in own chunk and previous chunk, but we attend ALL bucket's attention score 

since self.causal = False by default.

Am i missing something?

**Comment by lucidrains** [comment]
@muiPomeranian each chunk is meant to approximate the length of a bucket, assuming all query/keys are distributed uniformly to all buckets. the authors make the assumption that the buckets will rarely exceed the length of two chunks, which is why you have one chunk attending to itself and the one previous

**Comment by muiPomeranian** [comment]
Also, paper said they reduced/limited the computation of self attention(at LSHAttention class forward) but the way we implement is (since you followed the same way as the original trax code) that we calculate all self attention first. I think I am missing something here ..!!

so I was thinking, 

1. self.causal should be True,
2. this if self.causal part should be coming before we do einsum for self attention score. 

please excuse me if I am doing something very wrong !

**Comment by lucidrains** [comment]
@muiPomeranian yea, I think you may misunderstand how it works. attention comes in two steps, measuring the similarity of the query/keys, and then a softmax, which then yields the attention (you are forcing the network to pay attention to which query/key interaction is most important). masking happens between these two operations, to redirect attention to only those query/keys you want.

**Comment by muiPomeranian** [comment]
> @muiPomeranian each chunk is meant to approximate the length of a bucket, assuming all query/keys are distributed uniformly to all buckets. the authors make the assumption that the buckets will rarely exceed the length of two chunks, which is why you have one chunk attending to itself and the one previous

yeap! I understood from your previous response that that's why we expand K to [16,256,64,128]. 
with look_one_back ! 

but, my question is a little different. 
for example,
We calculate all self-attention with Q0 with K0~K127. 
and then

if self.causal = True, we Mask with max negative value for those who are not in same bucket.(e.g. Mask self attention value from K30, K122 if K30 and K122 are not in same bucket with Q0)

but this self.causal=False, and this is after einsum

**Comment by lucidrains** [comment]
@muiPomeranian `causal` has nothing to do with attention across buckets. it has to do with preventing the past from attending to the future, for training an auto-regressive model

**Comment by muiPomeranian** [comment]
I see, my bad... then !

https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L318

this part is we do calc the self attention activation and seems we are calculating all Q and K's regardless of they are in same bucket or not. am i right? 

we do get dot products of 

Q=[16,256,64,64]
K=[16,256,64,128] <- expanded by look_one_back

we can't say that all Q and K are in same bucket. aren't we?

**Comment by muiPomeranian** [comment]
> st from attending to the future, for tr

oh you are very right ..... 

https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L341

so this line, we prevent attending the future(right side part). 

then, this should be True aren't we? that's what paper recommended.. we only attend past bucket(previous bucket's).

**Comment by lucidrains** [comment]
@muiPomeranian so yea, that part is only the first step of attention. the attention procedure is completed here https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L394

yup, we are measuring the similarity of QKs that are within one chunk and the one before.

**Comment by muiPomeranian** [comment]
> @muiPomeranian so yea, that part is only the first step of attention. the attention procedure is completed here https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L394
> 
> yup, we are measuring the similarity of QKs that are within one chunk and the one before.

then are we assuming that all chunk and the one before are having same buckets?? (like blue color for instance ....according to paper)

**Comment by lucidrains** [comment]
@muiPomeranian the buckets have nothing to do with the order of the sequence, they are simply bucketing all similar tokens by LSH

**Comment by lucidrains** [comment]
@muiPomeranian yup, it is an assumption, may not necessarily be true. (actually, a researcher friend and I found that LSH didn't really yield as uniform a bucket distribution as reported in the paper)

**Comment by lucidrains** [comment]
@muiPomeranian you can examine both the attention and bucketing with the tools I provide in this section https://github.com/lucidrains/reformer-pytorch#research

**Comment by muiPomeranian** [comment]
wait ! so just to clarify ,

let's say im in Q=[16,256,64,64], K=[16,256,64,128].

paper says we can assume
Q[0,0,:,:][0] and L[0,0,:,:][0] are all in same bucket

so we skip this part:
https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L354

but you found out that it is not true, there might be a different bucket residing in Q[0,0,:,:][0] and L[0,0,:,:][0]

am i right?

(btw thanks a lot)

**Comment by lucidrains** [comment]
@muiPomeranian exactly! the sort / chunking is a crude way of maximizing the amount of interactions of QKs that fall into the same bucket. but depending on the actual bucket distribution, you could miss out on valuable QK interactions (if the length of the bucket exceeds twice the chunk length)

**Comment by lucidrains** [comment]
@muiPomeranian the most accurate way is to sort all tokens by buckets, find the length of the longest bucket, and then organize the tokens into a tensor with that longest bucket length, the rest of the buckets padded. maybe you could figure out how to do that efficiently, if you are part of a lab investigating this

**Comment by lucidrains** [comment]
@muiPomeranian i'm guessing they probably went with the rough chunking method because it is good enough

**Comment by muiPomeranian** [comment]
oh wow, can't believe google made this assumption in code ! 
Q1)
so to actually follow their instruction, we should NOT skip this part. 
but they skipped it since they believe their assumption is true.

https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L354

Q2)
so their paper results will still not considering this:
https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L354
(which means they still skip this line since they have wrong assumption that all buckets are sorted )

Q3) 
can you elaborate this a little more specifically? im sorry ...

_the most accurate way is to sort all tokens by buckets, find the length of the longest bucket, and then organize the tokens into a tensor with that longest bucket length, the rest of the buckets padded. maybe you could figure out how to do that efficiently, if you are part of a lab investigating this_

I thought buckets are following the original input order of characters(already sorted). I thought only part that original code missing is self.causal=False..

**Comment by lucidrains** [comment]
if you set `attend_across_buckets = False`, then you would only be missing out on interactions of QKs that fall into buckets that exceed 2 chunks

**Comment by lucidrains** [comment]
@muiPomeranian I've asked the authors about that setting, and they say they haven't found a big difference if that were on or off. It makes sense, since we are more worried about missing highly similar QKs, which is covered by the `n_hashes`

**Comment by muiPomeranian** [comment]
oh wait, I think it got your point. 

1)
it's not necessarily true that we only need to do look_one_back once. we might need to have additional look_one_back on K, so like[16,256,64,64*M], M>2. 

2) This is what you meant by saying 

**sort all tokens by buckets, find the length of the longest bucket, and then organize the tokens into a tensor with that longest bucket length, the rest of the buckets padded.**

So, I **don't need to sort,** 
but need to find what the max M value should be !
Then, rest of the buckets who need less than M, just pad them.



2)
but author says this is negligible. only M=2 is enough since in worst case we will miss NOT highly similar QKs.

3)
https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L354
If this is skipped(which we are), we don't IGNORE the QKs from the other bucket. But this is negligible? 

4)
causal = False. So we **DO NOT** preventing the past from attending to the future, for training an auto-regressive model. 

unlike what paper described in their picture, we do attend to left and right of buckets! and again, this was negligible..?

**Comment by lucidrains** [comment]
@muiPomeranian yea, you are right, you could dynamically adjust the lookback M based on the maximum size of the set of buckets. great idea! please share your results with me once you do :)

yea, it is negligible. think of it this way. you are organizing a speed dating event for a large number of people, and you want to maximize the number of final couples. you group all the people into buckets by some similarity measure (in this case, LSH). even if you accidentally have some people pair off across buckets by accident, if the LSH holds, it should be less of an attraction (attention) that it will be filtered out by the final softmax anyways.

**Comment by lucidrains** [comment]
in the worst case, you have every person pair off with every other person (assume these people are androgynous), and it becomes full attention.

**Comment by muiPomeranian** [comment]
cool ! may you confirm that 3) and 4) are correct too? If so, I might just want to see details ! 

+
yeah will share my result too after get the loss :)

**Comment by muiPomeranian** [comment]
3) even though paper suggest we should attend the QK in same bucket, they might miss(by not expanding it more than M=2) or even include(irrelvant bucket K)

4) even though paper suggest we should only attend the past bucket(previous one or left side), the code is attending ALL

and 3) 4) are ignored for their experiment and our code here since the result change is negligible

**Comment by lucidrains** [comment]
3) yes, exactly. i think the miss is more costly than including. but remember, they do `n_hashes` here, to reduce the number of misses, and it seems to work well.

4) the lookback is simply to reduce the misses, it has nothing to do with past/future and has nothing to do with the `causal` flag

**Comment by lucidrains** [comment]
@muiPomeranian i'm hoping to clear up any confusion around `causal`. The causal flag, when turned on, masks out any QK's trying to attend to QK's of the future. This is done independent of the bucketing, chunking, lookback

**Comment by muiPomeranian** [comment]
Oh yes! For 4), its nothing related to bucket, lookback. Its only for the causal.

What i meant for 4) was

Paper said causal=true(we only attend previous bucket) but actual implementation was causal=false(we attend both left and right!!)

**Comment by lucidrains** [comment]
@muiPomeranian yup, the paper did experiments with causal turned on, for building a generative model. but you could also use this for BERT-like masked-language-modeling with `causal=False`

**Comment by lucidrains** [comment]
wheter `causal` is on or off means it still attends within the chunk and its lookback chunk, just so you know. it's just with causal turned on, you filter our past attending to future within that chunk.
